docs/05-actor-model/idea.md

```diff
@@ -0,0 +1,327 @@
+# Idea behind an Actor Model
+
+Actor model is the solution to the problem of communication between smart
+contracts. Lets take a look why this particular solution is chosen in
+CosmWasm, and what are the consequences of that.
+
+## The problem
+
+Smart contracts can be imagined as sandboxed microservices. Due to SOLID
+principles it is valuable to split responsibilities between entities.
+However to split the work between contracts themselves, there is a need
+to communicate between them, so if one contract is responsible for
+management group membership, it is possible to call its functionality
+from other contract.
+
+The traditional way to solve this problem in SW engineering is to
+model services as functions which would be called with some RPC
+mechanism, and return its result as a response. As this approach looks
+nice, it creates sort of problems, in particular with shared state
+consistency.
+
+The other approach which is far more popular in business level modeling
+is to treat entities as actors, which can perform some task, but without
+interrupting it with calls to another contracts. Any calls to another
+contracts can be called only after whole execution is performed. When
+"subcall" would be finished, it will call the original contract back.
+
+This solution may feel unnatural, and it actually requires to kind of
+different design solutions, but it turns out working pretty good for
+smart contracts execution. In this section I will try to step by step
```

Comment:
I will try to explain how to reason about it, and how it maps to contract structure step by step.